espnet2/train/checkpoint.py

```diff
@@ -0,0 +1,160 @@
+        latest = None
+        for p in output_dir.glob("*epoch/timestamp"):
+            try:
+                n = int(p.parent.name.replace("epoch", ""))
+            except TypeError:
+                continue
+            with p.open("r") as f:
+                # Read the timestamp and comparing
+                date = f.read().strip()
+                try:
+                    date = datetime.strptime(date, "%Y-%m-%d %H:%M:%S")
+                except ValueError:
+                    continue
+            if latest is None or date > latest:
+                resume_epoch = n
+                latest = date
+
+    # If not found any snapshots, then nothing is done
+    if resume_epoch == 0:
+        resume_epoch = None
+
+    if resume_epoch is not None or resume_path is not None:
+        if resume_path is None:
+            resume_path = output_dir / f"{resume_epoch}epoch"
+            logging.info(f"--resume_epoch {resume_epoch}: Loading from {resume_path}")
+        else:
+            logging.info(f"--resume_path {resume_path}: Loading from {resume_path}")
+        load_checkpoint(
+            resume_path=resume_path,
+            model=model,
+            reporter=reporter,
+            optimizers=optimizers,
+            schedulers=schedulers,
+            map_location=map_location,
+        )
+
+
+def load_checkpoint(
+    resume_path: Union[str, Path],
+    model: torch.nn.Module,
+    reporter: Reporter,
+    optimizers: Iterable[torch.optim.Optimizer],
+    schedulers: Iterable[Optional[AbsScheduler]],
+    map_location: str,
+):
+    for key, obj in [("model", model), ("reporter", reporter)]:
+        _st = torch.load(resume_path / f"{key}.pth", map_location=map_location)
+        if obj is not None:
+            obj.load_state_dict(_st)
+    states = torch.load(resume_path / f"optim.pth")
+    for o, s, state in zip(optimizers, schedulers, states):
+        o.load_state_dict(state["optim"])
+        if s is not None:
+            s.load_state_dict(state["scheduler"])
+
+
+def save_checkpoint(
+    save_path: Union[str, Path],
+    model: torch.nn.Module,
+    reporter: Reporter,
+    optimizers: Iterable[torch.optim.Optimizer],
+    schedulers: Iterable[Optional[AbsScheduler]],
+):
+    for key, obj in [("model", model), ("reporter", reporter)]:
+        save_path.mkdir(parents=True, exist_ok=True)
+        p = save_path / f"{key}.pth"
+        p.parent.mkdir(parents=True, exist_ok=True)
+        torch.save(obj.state_dict() if obj is not None else None, p)
+    states = tuple(
+        {
+            "optim": o.state_dict(),
+            "scheduler": s.state_dict() if s is not None else None,
+        }
+        for o, s in zip(optimizers, schedulers)
+    )
+    torch.save(states, save_path / f"optim.pth")
+
+    # Write the datetime in "timestamp"
+    with (save_path / "timestamp").open("w") as f:
+        dt = datetime.now()
+        f.write(dt.strftime("%Y-%m-%d %H:%M:%S") + "\n")
+
+
+def load_pretrained_model(
```

Comment:
better to add comments to clarify the distinction between `load_pretrained_model` and `load_checkpoint`?